lib/inspec/utils/telemetry/run_context_probe.rb

```diff
@@ -0,0 +1,45 @@
+module Inspec
+  module Telemetry
+    # Guesses the run context of InSpec - how were we invoked?
+    # All stack values here are determined experimentally
+
+    class RunContextProbe
+      def self.guess_run_context
+        stack = caller_locations(4)
+        return "test-kitchen" if kitchen?(stack)
+        return "cli" if run_by_thor?(stack)
+        return "audit-cookbook" if audit_cookbook?(stack)
+        "unknown"
+      end
+
+      def self.run_by_thor?(stack)
+        stack_match(stack: stack[-14..-9], path: "thor/command", label: "run") &&
+        stack_match(stack: stack[-14..-9], path: "thor/invocation", label: "invoke_command")
+      end
+
+      def self.kitchen?(stack)
+        stack_match(stack: stack[-21..-11], path: "kitchen/instance", label: "verify_action") &&
+        stack_match(stack: stack[-14..-4], path: "kitchen/instance", label: "verify")
+      end
+
+      def self.audit_cookbook?(stack)
+        stack_match(stack: stack[-21..-11], path: "chef/handler", label: "run_report_handlers") &&
+        stack_match(stack: stack[-26..-16], path: "handler/audit_report", label: "report")
```

Comment:
These fixed numbers like `-26` feel likely to change and break, is that true? What happens if they do? What are the consequences of false positives/negatives?

At the least I think we'd want unit tests that run against a captured stack trace embedded in the tests. That would make it easier to understand what is being picked out by seeing the same. Of course that wouldn't detect any changes to the stack, so we'd probably want an integration tests for each of these, no?

Reply:
I've removed the fixed frame ranges.  I agree they would be brittle. 

The consequence of a false positive would be miscategorizing a run in the telemetry data. A false negative would likely result in an "unknown" result.

I've added integration tests that exercise each of the three outcomes.